velox/docs/programming-guide/chapter01.rst

```diff
@@ -0,0 +1,376 @@
+bytes, which is 13. The last byte has some bits unused.
+
+We have learned how to allocate memory and fill it in with values and null
+flags. We are now ready to assemble a flat vector to hold data for a single
+column. Let’s make a vector to store 100 sequential BIGINT values with every
+other value being null: [0, null, 2, null, 4, null, 6,..].
+
+.. code-block:: c++
+
+    #include "velox/vector/FlatVector.h"
+
+    auto vector = std::make_shared<FlatVector<int64_t>>(
+       pool.get(), BIGINT(), nulls, 100, values, std::vector<BufferPtr>{});
+
+    LOG(INFO) << vector->toString();
+
+    > [FLAT BIGINT: 100 elements, 50 nulls]
+
+To make a vector, we use FlatVector<T> class template with T being int64_t
+(64-bit integer). We pass a pointer to MemoryPool, a Type object that describes
+the type of values to store, nulls buffer, number of values, values buffer and
+an empty list of string buffers. Let’s ignore the string buffers for now. We’ll
+discuss these later.
+
+FlatValue<T> class can be used to store values of primitive types. The following
+types are supported in Velox:
+
+======================  ===========================    ==================
+Type                    C++ Type                       Description
+======================  ===========================    ==================
+BOOLEAN                 bool                           A boolean flag: true or false.
+TINYINT                 int8_t                         8-bit integer.
+SMALLINT                int16_t                        16-bit integer.
+INTEGER                 int32_t	                       32-bit integer.
+BIGINT                  int64_t                        64-bit integer.
+REAL                    float                          32-bit floating point number.
+DOUBLE                  double                         64-bit floating point number.
+VARCHAR                 struct StringView              Variable width string.
+======================  ===========================    ==================
+
+Nulls buffer can be null, which indicates that all values are not null.
+
+.. code-block:: c++
+
+    auto nonNullVector = std::make_shared<FlatVector<int64_t>>(
+       pool.get(), BIGINT(), nullptr, 100, values, std::vector<BufferPtr>{});
+
+    LOG(INFO) << nonNullVector->toString();
+
+    > [FLAT BIGINT: 100 elements, no nulls]
+
+Values buffer can also be null in case all values are null.
+
+.. code-block:: c++
+
+    auto nulls = allocateNulls(100, pool.get(), bits::kNull);
+    auto allNullVector = std::make_shared<FlatVector<int64_t>>(
+       pool.get(), BIGINT(), nulls, 100, nullptr, std::vector<BufferPtr>{});
+
+    LOG(INFO) << allNullVector->toString();
+
+    > [FLAT BIGINT: 100 elements, 100 nulls]
+
+However, it is invalid to create a vector with both nulls and values buffers being null.
+
+.. code-block:: c++
+
+    std::make_shared<FlatVector<int64_t>>(
+       pool.get(), BIGINT(), nullptr, 100, nullptr, std::vector<BufferPtr>{});
```

Comment:
Would be great to point out the relation between the C++ type and Velox type here.